lib/improver/utilities/cube_units.py

```diff
@@ -32,14 +32,126 @@
+                [coord.name() for coord in cube.coords()])
+        all_coords = list(set(all_coords))
+    # modify the copied cubes in place
+    _enforce_coordinate_units_and_dtypes(new_cubes, all_coords)
+
+    if not enforce:
+        # check each cube against its counterpart and fail if changed
+        for cube, ref in zip(new_cubes, cubes):
+            if cube.units != ref.units:
+                msg = ('Units {} of {} cube do not conform to '
+                       'expected standard {}')
+                raise ValueError(msg.format(ref.units, ref.name(), cube.units))
+            if cube.dtype != ref.dtype:
+                msg = ('Datatype {} of {} cube does not conform to '
+                       'expected standard {}')
+                raise ValueError(msg.format(ref.dtype, ref.name(), cube.dtype))
+            for coord in all_coords:
+                try:
+                    if cube.coord(coord).units != ref.coord(coord).units:
+                        msg = ('Units {} of coordinate {} on {} cube do not '
+                               'conform to expected standard {}')
+                        raise ValueError(msg.format(
+                            ref.coord(coord).units, ref.coord(coord).name(),
+                            ref.name(), cube.coord(coord).units))
+                    if cube.coord(coord).dtype != ref.coord(coord).dtype:
+                        msg = ('Datatype {} of coordinate {} on {} cube does '
+                               'not conform to expected standard {}')
+                        raise ValueError(msg.format(
+                            ref.coord(coord).dtype, ref.coord(coord).name(),
+                            ref.name(), cube.coord(coord).dtype))
+                except CoordinateNotFoundError:
+                    pass
+
+    return iris.cube.CubeList(new_cubes)
+
+
+def _find_dict_key(input_key, error_msg):
+    """
+    If input_key is not in the DEFAULT_UNITS dict, test for substrings of
+    input_key that are available.  This allows, for example, use of
+    "temperature" and "probability" in the DEFAULT_UNITS dict to avoid multiple
+    duplicate entries.
+
+    Args:
+        input_key (str):
+            Key that didn't return an entry in DEFAULT_UNITS
+        error_msg (str):
+            Error to raise if no unique match is found
+
+    Returns:
+        str: New key to identify required entry
+
+    Raises:
+        KeyError: If the function finds either zero or multiple matches
+
+    """
+    if "probability" in input_key:
+        # this avoids duplicate results from key matching below
+        return "probability"
+
+    matching_keys = []
+    for key in DEFAULT_UNITS.keys():
+        if key in input_key:
+            matching_keys.append(key)
+    if len(matching_keys) != 1:
+        raise KeyError(error_msg, 'matching keys:', matching_keys)
+
+    return matching_keys[0]
+
+
+def _enforce_coordinate_units_and_dtypes(cubes, coordinates, inplace=True):
```

Comment:
Should we consolidate these two functions (coordinate and diagnostic)? I probably should have originally, but now they are nicely wrapped and private I think it is even simpler to remove the duplication, e.g. keep the two function names we have here to basically set up the loops, but everything inside the for loops appears to be (pretty much) common and could be a standalone function; pass in the feature of the cube we want to check/modify, either the cube (for diagnostic data) or a coordinate, and the relevant variable name. Any thoughts? It might get rid of a fair bit of duplication.

Reply:
Yes.  I rather wondered whether this might be a plugin, actually.  There are a few ways this could be done - let's discuss.